lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1015,21 +1015,30 @@ NullablePtr<Pattern> TypeChecker::trySimplifyExprPattern(ExprPattern *EP,
                                                          Type patternTy) {
   auto *subExpr = EP->getSubExpr();
   auto &ctx = EP->getDeclContext()->getASTContext();
+  const auto optType = patternTy->getOptionalObjectType();
 
-  if (patternTy->isBool()) {
-    // The type is Bool.
+  if (patternTy->isBool() || (optType && optType->isBool())) {
+    // The type is Bool or Optional<Bool>.
```

Comment:
IMO this ought to be able to handle any level of optionality, you can either use `lookThroughAllOptionalTypes` and re-apply the same level of wrapping after, or you could refactor this function to be recursive and have it try unwrapping a single level as one of the checks(you'd want to handle the `nil` case before that though)